lib/src/keys.dart

```diff
@@ -0,0 +1,212 @@
+  ///
+  /// This should be paired with [p256dh].
+  String get auth => base64UrlEncode(_authKey);
+
+  /// Get the public key in webcrypto format.
+  Future<EcdhPublicKey> get pubKey =>
+      EcdhPublicKey.importRawKey(_pubKey, curve);
+
+  /// Get the raw authentication key.
+  Uint8List get authKey => _authKey;
+
+  WebPublicPushKey({
+    required List<int> publicKeyBytes,
+    required Uint8List authKeyBytes,
+  })  : _pubKey = publicKeyBytes,
+        _authKey = authKeyBytes;
+
+  /// Validate the key parameters.
+  Future<WebPublicPushKey> validate() async {
+    try {
+      if (_authKey.length != 16) {
+        throw KeyError('Invalid authentication key length ${_authKey.length}',
+            'authKeyBytes');
+      }
+      await pubKey;
+    } catch (e) {
+      if (e is FormatException && e.message.contains("INVALID_ENCODING")) {
+        throw KeyError('Invalid public key', 'publicKeyBytes');
+      }
+      rethrow;
+    }
+    return this;
+  }
+
+  /// For later use with [WebPublicPushKey.fromMap].
+  Map<String, List<int>> toMap() => {'p256dh': _pubKey, 'auth': _authKey};
+
+  /// Serializes public key for storage.
+  ///
+  /// Deserialize with [WebPushKeys.deserialize].
+  String get serialize => '$p256dh+$auth';
+
+  static Future<WebPublicPushKey> fromMap(Map<String, List<int>> keys) async {
+    if (keys case {'p256dh': List<int> pub, 'auth': List<int> auth}) {
+      return WebPublicPushKey(
+        publicKeyBytes: pub,
+        authKeyBytes: Uint8List.fromList(auth),
+      ).validate();
+    }
+    throw KeyError('bad public key/auth');
+  }
+}
+
+class WebPrivatePushKey {
+  final List<int> _privKey;
+
+  ///Get key in webcrypto format
+  Future<EcdhPrivateKey> get privKey =>
+      EcdhPrivateKey.importPkcs8Key(_privKey, curve);
+
+  WebPrivatePushKey({required List<int> privateKeyBytes})
+      : _privKey = privateKeyBytes;
+
+  /// Validate the key parameters.
+  Future<WebPrivatePushKey> validate() async {
+    try {
+      await privKey;
+    } catch (e) {
+      if (e is FormatException && e.message.contains("INVALID_ENCODING")) {
+        throw KeyError('Invalid private key', 'privateKeyBytes');
+      }
+      rethrow;
+    }
+    return this;
+  }
+
+  /// For later use with [WebPrivatePushKey.fromMap].
+  Map<String, List<int>> toMap() => {'priv': _privKey};
+
+  /// Serializes public key for storage.
+  ///
+  /// Deserialize with [WebPushKeys.deserialize].
+  String get serialize => base64Url.encode(_privKey);
+
+  static Future<WebPrivatePushKey> fromMap(Map<String, List<int>> keys) async {
+    if (keys case {'priv': List<int> privateKeyBytes}) {
+      return WebPrivatePushKey(
+        privateKeyBytes: privateKeyBytes,
+      ).validate();
+    }
+    throw KeyError('bad public key/auth');
+  }
+}
+
+/// Stores the public, private, auth key needed for webpush.
+class WebPushKeys {
+  final WebPublicPushKey? publicKey;
+  final WebPrivatePushKey? privateKey;
```

Comment:
```suggestion
  final PublicWebPushKey publicKey;
  final PrivateWebPushKey privateKey;
```

I suggest to:
- Rename the variables so "WebPush" is not splitted. Public and Private are set in the beginning to make it clearer.
- Force having both. A key set is a pubkey + privkey